### DISCUSSION TITLE
Developer Operations

### DISCUSSION THREAD

**Discussion by arneschreuder** [COS-301/graduates#15] [General]
This is the discussion board for and with Developer Operations.

**Comment by arneschreuder** [comment]
Please follow: https://nx.dev/ci/monorepo-ci-github-actions

**Comment by arulagrawal** [comment]
Hi, we have some questions related to general git workflow.

My current understanding is this:

- Everyone in the team forks COS-301/graduates
- Everyone makes the branch feature/team/feature_name
- Each person commits their work to their own fork on the branch feature/team/feature_name.
- DevOps merges everyone's feature branch with upstream feature branch (via pull requests)
- Either DevOps or Project Owners merge from feature branch to develop.

Is that the general idea?

**Reply by LiamBurgess18015001** [reply]
@arneschreuder

**Reply by arneschreuder** [reply]
Hi There. That sounds almost right. Let me correct where wrong:

* Everyone, like literally everyone forks the repo. No work will be done on upstream unless through PR.
* Everyone, like literally everyone that works on anything related to code, makes feature branches using the convention `feature/{whatever}`, no team, nothing like that in the name.
* Correct
* DevOps is not responsible for this, this is the PMs job, but you can certainly assist. However, you need to keep an eye on ALL PRs, since some PRs might break the pipeline or conventions are not correct etc, then you need to step in. If a PR breaks the pipeline, this team, and the DevOps from that team must figure it out. To be specific PRs are made from `arneschreuder:feature/whatever` to `upstream:develop`.
* This is the PMs and the person who made the PR's responsibility, DevOps just participates if there are maybe things wrong, missing etc.


I will show more examples of this on Friday.

**Reply by arneschreuder** [reply]
The key cultural thing to get right here is that:

* PRs should be MANY and should be SMALL. Not few and large. So collaboration happens through getting stuff in develop, not through working on feature branches together. If anybody is working on a branch together, that means they are missing the point and you can help mitigate this.
* In the beginning, develop might break a lot, since there is a lot of dependencies between teams etc. But those should be catered for using mocks. If I am waiting for the auth team, for some service that they will have, they should've already provided some mock service that I can at least temporarily use. So allow for mistakes in the beginning ok, people will fall into the motion of things.

**Reply by arulagrawal** [reply]
Thanks, that clarifies quite a lot! Since things like tests aren't setup yet (and therefore no automated testing via CI) what are some of the things DevOps can do to provide value to the team in the beginning?

I remember you talking about gitmoji, would for example a check to ensure commit messages follow gitmoji be something doable? Or is that more just on the side?

**Reply by arulagrawal** [reply]
@arneschreuder

**Reply by arneschreuder** [reply]
Sorry, I see I never replied here @arulagrawal . Apologies. DevOps needs to work with CI/CD team to set up the automated testing ASAP. That should be one of the first commits. Even if there are no tests (which there should be some tests already afaik). The pipeline is agnostic whether the implementation is there of not. It just provides the plumbing. So I would say that is an urgent todo. Those repo chores that you mention are also important, but those only come after CI/CD has implemented their part. The DevOps person needs to take this pipeline to the team, and make sure the team is prepping for it: by having tests, linted, etc. Look at like Husky, to do precommit hooks.

**Comment by DYL-WF** [comment]
@Alexander-Muendesi, @COS-301/go from what I have gathered your team is responsible for CI/CD. Please let me know if you need anything from my end as a devOP. 

I eagerly await any updates from your end. At the moment, your wiki is empty.

**Reply by Alexander-Muendesi** [reply]
Hey there.

We are still in the phase of "System Requirements Definition" and hope to be done early next week. Once that is done the wiki will be updated accordingly and we will coordinate with DevOps members.

In the mean time did you perhaps visit the conversation I had with Arne  about the  whole CI/CD and devOps thing in the Project Manager discussion board? That could clear things up a bit if anything is foggy

Link: https://github.com/COS-301/graduates/discussions/3#discussioncomment-2319071

**Reply by DYL-WF** [reply]
@Alexander-Muendesi is there a WhatsApp group? if so can I please get a link to join it? I'm sorry, I missed the first meeting.

**Reply by Alexander-Muendesi** [reply]
There is no group that I am aware of. Maybe ask the other DevOps people. I dont know why there is a need for one though? My team will be adding some scripts most probably by the end of Sunday to the pipeline.  

Also keep in mind I am a PM for CI/CD, not DevOps. Maybe if you guys have a lead devOps rather message that individual

**Comment by 10416260** [comment]
@sinothandomabhena  Please take a look at this thread

**Reply by Alexander-Muendesi** [reply]
@sinothandomabhena if you have some specific questions just drop them here and I will answer them best I can. If you require a discord voice chat I  feel like you will gain more clarity talking to my DevOps member. He is more in tune with things

**Comment by 10416260** [comment]
@Alexander-Muendesi Please help by getting in touch with @sinothandomabhena , he would like to have more discussion when it comes to the DevOps role, he's actually a UI engineer but his role involves working closely with the DevOps team in terms of deployment

**Comment by arulagrawal** [comment]
@arneschreuder 

Should a DevOps member be providing team wide github actions? Or should github actions be left to the CI/CD team?

If I should provide some team wide automations then what is the flow to do so? 2 ideas were brought up in the DevOps meeting

1. I make the actions on my fork, and everyone forks my fork to have access to my actions.
2. I make the actions on my fork, and everyone PRs to my fork to get the actions to run. When the actions pass, we make PR to upstream.

Which of the 2 is preferable? Should I even worry about github actions as someone who is not on the CI/CD or DevOps team?

**Reply by arneschreuder** [reply]
No need for team-based actions. The global CI/CD actions is all we need. Your role is then to sure that you help get your team's code into develop. If your team's PRs break the development branch, I am going to come ask you. So you have to ensure that the team confirms to global CI/CD pipeline and developer processes.

**Reply by arulagrawal** [reply]
Okay so just to be sure, I would have like 0 commits since I am not providing code nor actions?
Also, I thought about writing a contribution guide in the project wiki. Something that overviews what a developer (or anyone contributing) should do to get their changes into upstream. Essentially an overview of the develop process. Is this a good idea? @arneschreuder

**Reply by arneschreuder** [reply]
Absolutely. 

Also, it doesn't necessarily mean you will have 0 commits. You might help you team to get their PRs in upstream, and that might require you jumping in to help.

**Reply by arulagrawal** [reply]
Perfect thanks

**Comment by MichaelViljoen** [comment]
@arneschreuder 
We understand that we need two docker files, one for the client and one for the API, is it appropriate for us to store them in the root, and do they need a specific name.

**Reply by arneschreuder** [reply]
They need to be in `apps/api/` and `apps/client/`

**Comment by nimGaluu** [comment]
@arneschreuder, 

Concern about friday's Demonstrations - in short:
       Since our major task for each team has already been allocated as a feature, CI/CD, and not all teams have had a lot of git movements that we might need to monitor or 'supervise'. According to the playbook, Implementations start in sprint 2 and I believe then any DevOps member would have things to do since that will be the time PRs would be sent in and a lot of pipeline maintenance would be needed.
       WIth that said, I am afraid not much is there to present since we have minimum to none contribution as a Role. Those who have double roles would have advantage over those who have chosen this role only. Is that fair?

Please advice

KInd Regards,
_en

**Reply by arneschreuder** [reply]
Hi There. Thanks for bringing this up. I suspect we might need to broadcast some clarity as to what is expected for Friday. But basically the project managers need to come show up what their team did in sprint 1. These could be meetings, discussions, contributions to the wiki, code, code reviews, etc. We will probably on our side pull the repo log to just verify. Nonetheless, you should think of sprint 1 evaluation as a small progress update. We will give a small mark (like 3 of the 10) for spring 1 evaluation and the only requirement is that you have deliberately worked towards your deliverable. 

So in the case of DevOps. You should have had meetings now, talked and discussed how CI/CD is going to be working with the CI/CD feature team. Ensure that everybody is coming right with the repo. If there are PRs, assistance here helps. Most importantly, if you have tasks allocated to you, you should be fine.

**Comment by DYL-WF** [comment]
So my API has just done somework on his end and has now made a pull request. Can someone please confirm with me that our process was correct. 

-> Updated Fork
-> Pull from fork

->created new branch(shorts-api) off develop -> added API files
-> commited
-> pushed to origin(Fork)

-> Made PR to upstream from FORK/feature-branch

**Reply by arneschreuder** [reply]
Sounds good

**Comment by arneschreuder** [comment]
Hi @COS-301/developer-operations you have your first challenge. So the CICD team added the basics of the GitHub actions last night. Well done to them. So the pipeline is currently broken. Now this is where as Developer Operations come in, we jump in, investigate the issue and help fix it. You will see it is mostly linting issues.

**Reply by arneschreuder** [reply]
See here: https://github.com/COS-301/graduates/actions

**Reply by arneschreuder** [reply]
Also, at some point, please take a look at: https://github.com/COS-301/graduates/community

**Reply by arulagrawal** [reply]
What will be the procedure to fix issues related to the test-all lint check failing? Our PR is failing because the test-all lint fails on code from other places.

Also @arneschreuder can I give your email when setting up a code of conduct?

**Reply by arulagrawal** [reply]
Nevermind about that first point I see https://github.com/COS-301/graduates/pull/280

**Reply by evilcomrade** [reply]
We should also put up a Pull request template (and maybe an Issue template as well), those seem pretty relevant.

**Reply by RobynHancock** [reply]
@arulagrawal Will you be compiling the code of conduct?

**Reply by arulagrawal** [reply]
I can

**Reply by RobynHancock** [reply]
@arulagrawal Okay. I can also help, if you want it

**Reply by arulagrawal** [reply]
@RobynHancock not gonna lie, I was gonna use one of the two templates offered by GitHub.

**Reply by arneschreuder** [reply]
@arulagrawal you can use mine sure. Yeah, this PR #284 fixed the pipeline. So now CI/CDs stuff is in and the rest of the PRs will be strict. People can not merge in if stuff fails.

**Comment by priyo5** [comment]
Hey guys 

Please take note that my team and I have added new tests to the repo 

Main purpose is to check meta tags of a PR

If one of these tests fail all you have to do is add the related tag it fails on

**Comment by arulagrawal** [comment]
Possible breakage of pipeline: 

I noticed PRs #498 and #500 failing with strange errors on the lint test.

I cloned upstream develop and ran the test locally:
`git clone https://github.com/COS-301/graduates`
`yarn install --frozen-lockfile`
`yarn nx run-many --target=lint --all`

and the result of this:
```yarn run v1.22.18
$ /tmp/main/graduates/node_modules/.bin/nx run-many --target=lint --all

Unexpected error:
TypeError: Cannot read properties of undefined (reading 'hashCommand')
    at /tmp/main/graduates/node_modules/@nrwl/linter/src/executors/eslint/hasher.js:11:32
    at Generator.next (<anonymous>)
    at /tmp/main/graduates/node_modules/tslib/tslib.js:117:75
    at new Promise (<anonymous>)
    at __awaiter (/tmp/main/graduates/node_modules/tslib/tslib.js:113:16)
    at run (/tmp/main/graduates/node_modules/@nrwl/linter/src/executors/eslint/hasher.js:7:34)
    at TasksSchedule.<anonymous> (/tmp/main/graduates/node_modules/nx/src/tasks-runner/tasks-schedule.js:117:19)
    at Generator.next (<anonymous>)
    at /tmp/main/graduates/node_modules/tslib/tslib.js:117:75
    at new Promise (<anonymous>)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.```

**Reply by arulagrawal** [reply]
@arneschreuder @COS-301/developer-operations @COS-301/go

**Reply by arulagrawal** [reply]
Is it possible there is some tooling change?
The test specified above fails as far back as 20 commits ago on the develop branch.

**Reply by DYL-WF** [reply]
yeah i was struggling a lot last night trying to find the issue.. Its all a little hazy but i think it has something to do with student-explore and the use of ng-packagr module but thats just a wild guess
.

**Reply by DYL-WF** [reply]
can someone please confirm whether the pipeline is broken or not please? cause i was planning on completely reseting  my commits and starting from scratch. Id hate to do that if the problem wasnt on my end this whole time. @arneschreuder @COS-301/developer-operations

**Reply by arulagrawal** [reply]
@DYL-WF, have a look at #502 and #504, Arne says he will look tonight

**Reply by DYL-WF** [reply]
@arulagrawal  should i hold of until tomorrow then for my pull request?

**Reply by DYL-WF** [reply]
but thanks. i was stressing that i broke something

**Reply by arulagrawal** [reply]
> @arulagrawal should i hold of until tomorrow then for my pull request?

holding off is probably best, any new PRs at this time would fail anyway

**Comment by Daniel-J-Burgess** [comment]
Hi everyone
I just want to confirm do we still need to link an issue to each PR?
It appears almost no one is doing this.

**Reply by Alexander-Muendesi** [reply]
you definitely should be doing that

**Comment by arneschreuder** [comment]
Hi @COS-301/developer-operations please fix for us. The post PR CI tests are failing, see error in actions:
<img width="1046" alt="Screenshot 2022-04-14 at 10 31 49" src="https://user-images.githubusercontent.com/1528314/163346585-8e55e5d0-419f-4747-9113-5304cad045dc.png">

**Reply by Alexander-Muendesi** [reply]
Hi @arneschreuder. With regards to that error, its is a very weird one.  The first time we tried implementing actions to run on merge, on our forks we never got such an error then when we moved it to repo it started giving that error and hence we(CI/CD) removed it. Then yesterday we did some changes to see what was causing it and made a PR to bring it back to the pipeline. It was working fine afterwards and all of a sudden from PR692 it starts complaining again. Sometimes it wants to work sometimes it does not want to work.

My team is struggling to see why its doing that. Any feedback would be appreciated

**Reply by arneschreuder** [reply]
Maybe it is the cache thing? Not sure. Maybe try removing the cache to see if that fixes it. That would be dumb, because then all the builds are going to take super long.

**Reply by DYL-WF** [reply]
should I hold all incoming PRs from my team until this issue is resolved @Alexander-Muendesi ?

**Reply by Alexander-Muendesi** [reply]
You can continue with them. We want to see with fix added if it will break again. The fix was just added now

**Comment by tshegomotlatle** [comment]
Could the team responsible for the student-explore page shed some light on this error that is currently experienced when the Full CI tests run
![Screenshot 2022-04-16 070548](https://user-images.githubusercontent.com/70272858/163662353-9bb30ba4-cbaa-4744-85e1-90b0eeca6a98.png)

**Reply by tshegomotlatle** [reply]
@COS-301/java

**Reply by Daniel-J-Burgess** [reply]
Appears to be caused by the recent database changes, we will update it.

**Reply by Daniel-J-Burgess** [reply]
@COS-301/sas  the company profile API will need to be updated as well.

**Reply by timovdmerwe** [reply]
Yes, the database has changed. I'll fix it asap

**Comment by MichaelViljoen** [comment]
@arneschreuder what is the turnaround time for reviewing pull requests?

**Reply by arneschreuder** [reply]
You should ask your own team… your own team should review and approve the PR.

**Reply by MichaelViljoen** [reply]
Sorry, will sort it out in my side.

**Comment by lindzik** [comment]
@arneschreuder for PRs that are now being made into release we need to have 2 approving reviews. Are you the second reviewer or can a team member review it for us?

**Reply by arneschreuder** [reply]
For now I will review it, but if it becomes much I will set it to one reviewer

**Comment by lindzik** [comment]
![e2e error](https://user-images.githubusercontent.com/94564461/164448473-c220528d-eb7b-400e-989f-2bdaa6c06aaf.PNG)

I got this error for the affected e2e tests for my team's PR. client shorts team do you have a fix for it.
